sync/device.ts

```diff
@@ -0,0 +1,177 @@
+import * as Docker from 'dockerode';
+import { Dockerfile } from 'livepush';
+import * as _ from 'lodash';
+import { Builder } from 'resin-docker-build';
+
+import { child_process, fs } from 'mz';
+import * as Path from 'path';
+import { Duplex, Readable } from 'stream';
+import * as tar from 'tar-stream';
+
+export function getDocker(deviceAddress: string): Docker {
+	return new Docker({
+		host: deviceAddress,
+		// TODO: Make this configurable
+		port: 2375,
+	});
+}
+
+export async function getSupervisorContainer(
+	docker: Docker,
+	requireRunning: boolean = false,
+): Promise<Docker.ContainerInfo> {
+	// First get the supervisor container id
+	const containers = await docker.listContainers({
+		filters: { name: ['resin_supervisor'] },
+		all: !requireRunning,
+	});
+
+	if (containers.length !== 1) {
+		throw new Error('supervisor container not found');
+	}
+	return containers[0];
+}
+
+export async function getDeviceArch(docker: Docker): Promise<string> {
+	try {
+		const supervisorContainer = await getSupervisorContainer(docker);
+		const arch = supervisorContainer.Labels?.['io.balena.architecture'];
+		if (arch == null) {
+			// We can try to inspect the image for the
+			// architecture if this fails
+			const match = /(amd64|i386|aarch64|armv7hf|rpi)/.exec(
+				supervisorContainer.Image,
+			);
+			if (match != null) {
+				return match[1];
+			}
+
+			throw new Error('supervisor container does not have architecture label');
+		}
+
+		return arch.trim();
+	} catch (e) {
+		throw new Error(
+			`Unable to get device architecture: ${e.message}.\nTry specifying the architecture with -a.`,
+		);
+	}
+}
+
+export async function getCacheFrom(docker: Docker): Promise<string[]> {
+	const container = await getSupervisorContainer(docker);
+	return [container.Image];
+}
+
+// perform the build and return the image id
+export async function performBuild(
+	docker: Docker,
+	dockerfile: Dockerfile,
+	dockerOpts: { [key: string]: any },
```

Comment:
Doesn't dockerode expose a type for this?

Reply:
Unfortunately not.

Reply:
You can get the types via
```suggestion
	dockerOpts: Parameters<typeof builder.createBuildStream>[0],
```